fix: support IPv4 and IPv6 in dev command

danez opened the pull request:
#### Summary

Fixes #4803 

This fixes support with node 17+ which defaults to IPv6 when resolving `localhost`

The idea in this PR is to listen for `[::1]:port` aswell as `127.0.0.1:port`. This way we do not rely on the name resolution of `localhost` and support IPv4 and IPv6. That way we now also support frameworks that in theory spin up their dev server with IPv6 (which I haven't seen yet).

I tested with astro (which needed some patches to make ipv6 in their dev server work) and it worked fine.

I have to say though that I wasn't able to reproduce the behavior that node resolves `localhost` to `::1`, on my machine it always resolves to `127.0.0.1`.

Is anyone able to reproduce the behavior?

anmonteiro reviewed (dismissed):
I can repro on master with `v18.5.0` but not in this branch. Thank you!

danez commented:
> I have a couple questions about edge case behavior here:
> 
> - Does `Promise.race` cancel promises that don't win the race?
>   - If it doesn't, we will be leaking e.g. a `::1` waiting promise that just keeps running forever

We should test this. i don't think it will cause any side effects and will simply timeout at some point.

anmonteiro commented:
I'm worried about this pathological case:

```js
const promise1 = new Promise((resolve, reject) => {
  setTimeout(resolve, 500, 'resolved');
});

const promise2 = new Promise((resolve, reject) => {
  setTimeout(reject, 100, 'threw');
});

Promise.race([promise1, promise2]).then((value) => {
  console.log(value);
}).catch(x => console.log('caught', x));
```

Running the code above will log `caught threw`, which is a problem if one of the promises throws before the other one resolves.

For this specific case it probably doesn't matter (barring any `wait-port` bugs), given that `wait-port` handles errors itself, and, as you mentioned, we can probably just wait until the other (pending) promise times out.

danez commented:
Yes, you are absolutely right. I mixed two different things here. For one I fixed the issue that people are having with node 17+, but I also added new support for IPv6 which I do not think we really need (yet) as no framework is using that by default.

So I reverted the new IPv6 support and only fixed the problem now.

I also created https://github.com/dwmkerr/wait-port/pull/82, but there is no need to wait for that, maybe it is better not to use name resolution anyway.

Thanks for all the feedback @anmonteiro

anmonteiro reviewed (approved):
Confirmed that this fix also works, just like the previous `Promise.race` one.
